src/commands/skip.rs

```diff
@@ -1,76 +1,39 @@
 use crate::{
-    guild::cache::GuildCacheMap,
-    strings::{NOTHING_IS_PLAYING, SKIPPED, SKIP_VOTE_EMOJI, SKIP_VOTE_MISSING, SKIP_VOTE_USER},
-    utils::{create_response, get_voice_channel_for_user},
+    strings::{NOTHING_IS_PLAYING, SKIPPED},
+    utils::create_response,
 };
 use serenity::{
-    client::Context,
-    model::{id::GuildId, interactions::application_command::ApplicationCommandInteraction},
-    prelude::{Mentionable, RwLock, SerenityError, TypeMap},
+    client::Context, model::interactions::application_command::ApplicationCommandInteraction,
+    prelude::SerenityError,
 };
-use std::{collections::HashSet, sync::Arc};
 
 pub async fn skip(
     ctx: &Context,
     interaction: &mut ApplicationCommandInteraction,
 ) -> Result<(), SerenityError> {
     let guild_id = interaction.guild_id.unwrap();
-    let bot_channel_id = get_voice_channel_for_user(
-        &ctx.cache.guild(guild_id).await.unwrap(),
-        &ctx.cache.current_user_id().await,
-    )
-    .unwrap();
     let manager = songbird::get(ctx).await.unwrap();
     let call = manager.get(guild_id).unwrap();
 
+    let args = interaction.data.options.clone();
+    let to_skip = match args.first() {
+        Some(arg) => arg.value.as_ref().unwrap().as_u64().unwrap() as usize,
+        None => 1,
+    };
+
     let handler = call.lock().await;
     let queue = handler.queue();
 
     if queue.is_empty() {
         return create_response(&ctx.http, interaction, NOTHING_IS_PLAYING).await;
-    }
-
-    let mut data = ctx.data.write().await;
-    let cache_map = data.get_mut::<GuildCacheMap>().unwrap();
-
-    let cache = cache_map.entry(guild_id).or_default();
-    cache.current_skip_votes.insert(interaction.user.id);
-
-    let guild_users = ctx.cache.guild(guild_id).await.unwrap().voice_states;
-    let channel_guild_users = guild_users
-        .into_values()
-        .filter(|v| v.channel_id.unwrap() == bot_channel_id);
-    let skip_threshold = channel_guild_users.count() / 2;
-
-    if cache.current_skip_votes.len() >= skip_threshold {
+    } else {
+        for _ in 1..to_skip {
+            queue.dequeue(1);
+        }
```

Comment:
Dangerous approach because if the user uses an index that exceeds the size of the queue, it loops without reason. But worse, if you input something stupid like 100M the bot crashes (most likely because it takes the handler and doesn't release it after it's done looping).

I'd suggest changing `to_skip` to be `min(input, queue_size)` and then calling `modify_queue` because it's more readable (and should be more performant?).

Reply:
Nope, `modify_queue` is not more performant, and definitely not more readable in my opinion, since dequeue actually calls `modify_queue`.

Either way, I tested with 1000 and it handled it fine, but I'll implement said limiter.

Reply:
> Nope, modify_queue is not more performant, and definitely not more readable in my opinion, since dequeue actually calls modify_queue.

The queue is actually a `VecDeque`, represented like `[tail]..unused..[head]` in memory. This means popping from the front is super quick. It's the recalculating of tail/head that was concerning me.

I was suggesting [draining](https://doc.rust-lang.org/src/alloc/collections/vec_deque/mod.rs.html#1249-1251) because it looks more optimized than repeatedly [removing](https://doc.rust-lang.org/src/alloc/collections/vec_deque/mod.rs.html#1854) (note Serenity calls this and not `pop_front` if the index is 1) and is undoubtedly more idiomatic.

However, I took the liberty of measuring both approaches but the differences are negligible and vary too much.

> Either way, I tested with 1000 and it handled it fine, but I'll implement said limiter.

Well yeah, 1000 is a breeze but 100M isn't.

Reply:
I think anyways the `to_skip` needs to be minimum limited, so as not to waste loops dequeueing stuff that does not exist plus now allowing an user to crash the both if he wants to.

Reply:
Didn't even notice you kept the skip afterwards. Given that you're removing from the middle, drain seems even more obvious now... Also why not incorporate the current track into that range (making sure you stop it beforehand)?

Reply:
> Also why not incorporate the current track into that range (making sure you stop it beforehand)?

This was because I wanted to make sure that the skip is almost unnoticeable, and removes the overhead that might be caused by the removal of the tracks in the queue before it.